libshpool/src/session_restore/mod.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use shpool_protocol::TtySize;
+use tracing::info;
+
+use crate::config::{self, SessionRestoreMode};
+
+// To prevent data getting dropped, we set this to be large, but we don't want
+// to use u16::MAX, since the vt100 crate eagerly fills in its rows, and doing
+// so is very memory intensive. The right fix is to get the vt100 crate to
+// lazily initialize its rows, but that is likely a bunch of work.
+const VTERM_WIDTH: u16 = 1024;
+
+pub trait SessionSpool {
+    /// Resizes the internal representation to new tty size.
+    fn resize(&mut self, size: TtySize);
+
+    /// Gets the current content as a byte buffer.
+    fn current_contents(&self) -> Vec<u8>;
```

Comment:
I think either the method name or the doccomment should tell us something about the data format of the vector returned by this routine. It isn't that obvious that this is going to give us terminal control codes to restore the screen, including a clear code to reset the screen from any state back to a known good state. Given that this isn't the full contents of the terminal, I think "restore_buffer" might be a better name. The docs should also mention that the number of lines restored will be determined by the session restore mode used to create the spool.